Add migration guide for buildpack api 0.6 (#331)
buildpacks/docs: content/docs/reference/spec/migration/buildpack-api-0.5-0.6.md

+++
title="Buildpack API 0.5 -> 0.6"
+++

<!--more-->

This guide is most relevant to buildpack authors.

> See the [spec release](https://github.com/buildpacks/spec/releases/tag/buildpack%2Fv0.6) for buildpack API 0.6 for the full list of changes and further details.

Reviewer: Need to release it only after the spec is released (the link is currently broken).

> ### Opt-in layer caching

Reviewer 2: Should this and the fact that the layer types have been moved under a types table be two separate entries?

Reviewer: I agree. It's also 2 separate entries in the spec: https://github.com/buildpacks/spec/releases/tag/buildpack%2Fv0.6.

Author: I don't know... the fact that it's two separate issues is an artifact of how the RFC unfolded (see comment: https://github.com/buildpacks/spec/pull/151#issuecomment-708631142).

Author: However I do think the paragraph below could be re-worded...taking a look.

Author: How does it look now?

Buildpacks must now explicitly opt-in to layer re-use by writing `launch`, `build`, or `cache` keys to a new `[types]` table in `<layers>/<layer>.toml` (note that these keys were removed from the top level). If buildpacks do not modify `<layers>/<layer>.toml`, the layer will not be re-used, even if the buildpack in the previous build set any of these keys to `true`.

A Bash buildpack could write something like the following to `<layers>/<layer>.toml` in order to cache a layer:

```bash
cat >> layer.toml <<EOL
[types]

> cache = true

Reviewer: ```suggestion
  cache = true
```
I think that the 2 spaces are necessary but I'm not 100% sure.
Anyway, I think it's clearer to have these spaces.

Author: After discussion we think the spaces are not necessary.

EOL
```

### Buildpacks contribute default process type

> Buildpacks may now contribute the default process type by writing `default = true` in the process type definition in `<layers>/launch.toml`. An individual buildpack may only specify one process type with `default = true`. The lifecycle will choose, from all buildpack-provided process types, the last process type with `default = true` as the buildpack-provided default. A user may override the buildpack-provided default process type by passing `-process-type` to the exporter. Note: to enable buildpacks implementing older Buildpack APIs to work with newer buildpacks, the lifecycle will assume that buildpacks on Buildpack API < 0.6 intended for `web` processes to be the default.

Reviewer 2: Should we add an example here?

Author: Thanks for the suggestion!

> Buildpacks may now contribute the default process type by writing `default = true` in the process type definition in `<layers>/launch.toml`. An individual buildpack may only specify one process type with `default = true`. The lifecycle will choose, from all buildpack-provided process types, the last process type with `default = true` as the buildpack-provided default. A user may override the buildpack-provided default process type by passing `-process-type` to the exporter. (Note: to enable buildpacks implementing older Buildpack APIs to work with newer buildpacks, the lifecycle will assume that buildpacks on Buildpack API < 0.6 intended for `web` processes to be the default.)

Reviewer: ```suggestion
Buildpacks may now contribute the default process type by writing `default = true` in the process type definition in `<layers>/launch.toml`. An individual buildpack may only specify one process type with `default = true`. The lifecycle will choose, from all buildpack-provided process types, the last process type with `default = true` as the buildpack-provided default. A user may override the buildpack-provided default process type by passing `-process-type` to the exporter. Note: to enable buildpacks implementing older Buildpack APIs to work with newer buildpacks, the lifecycle will assume that buildpacks on Buildpack API < 0.6 intended for `web` processes to be the default.
```

> 

Reviewer 2: ```suggestion
For example - 

```toml
# launch.toml
[[processes]]
type = "<process type>"
command = "<command>"
args = ["<arguments>"]
direct = false
# the default field is new
default = false
```
```

### New fields in buildpack descriptor

To introduce more information into the Buildpack Registry search API, the following are added to `buildpack.toml`:
* A `description` free text field
* A `keywords` array
* A `licenses` array of table
Each license in `[[buildpack.licenses]]` can contain a `type` and `uri`, where `type` may use the [SPDX 2.1 license expression](https://spdx.org/spdx-specification-21-web-version), but it is not limited to identifiers in the [SPDX Licenses List](https://spdx.org/licenses/). If the buildpack is using a nonstandard license, then the `uri` key may be specified in lieu of or in addition to `type` to point to the license.

### exec.d on Windows
Buildpack API 0.5 introduced support for `exec.d` executables that may be used to modify the environment for an app process (similar to profile.d) but do not depend on a shell.
The launcher will execute each file in each `<layers>/<layer>/exec.d` directory and each file in each `<layers>/<layer>/exec.d/<process>` directory before the app process is started.

> Previously, this was only implemented on Linux, but it is now implemented on Windows as well. On Windows, the launcher will pass each `exec.d` executable, via the `CNB_EXEC_D_HANDLE` environment variable, the hex representation of an open file handle where the executable is expected to write its output. The output may contain any number of top-level key/value pairs, which will be sourced by the launcher in the app process's environment.

Reviewer: ```suggestion
Previously, this was only implemented on Linux, but starting from Buildpack API 0.6, it is implemented on Windows as well. On Windows, the launcher will pass each `exec.d` executable, via the `CNB_EXEC_D_HANDLE` environment variable, the hex representation of an open file handle where the executable is expected to write its output. The output may contain any number of top-level key-value pairs, which will be sourced by the launcher in the app process's environment.
```

I think that the second sentence (starting with "On Windows,") in this paragraph is too long. Can we break it to two sentences?

A Bash buildpack could write something like the following to `<layers>/<layer>.toml` in order to cache a layer:

```bash
cat >> layer.toml <<EOL
[types]
cache = true
EOL
```

### Buildpacks contribute default process type

Buildpacks may now contribute the default process type by writing `default = true` in the process type definition in `<layers>/launch.toml`. An individual buildpack may only specify one process type with `default = true`. The lifecycle will choose, from all buildpack-provided process types, the last process type with `default = true` as the buildpack-provided default. A user may override the buildpack-provided default process type by passing `-process-type` to the exporter. Note: to enable buildpacks implementing older Buildpack APIs to work with newer buildpacks, the lifecycle will assume that buildpacks on Buildpack API < 0.6 intended for `web` processes to be the default.
For example - 

```toml
# launch.toml
[[processes]]
type = "<process type>"
command = "<command>"
args = ["<arguments>"]
direct = false
# the default field is new

> default = false

Author: ```suggestion
default = false
```
```

+++
title="Buildpack API 0.5 -> 0.6"
+++

<!--more-->

This guide is most relevant to buildpack authors.

See the [spec release](https://github.com/buildpacks/spec/releases/tag/buildpack%2Fv0.6) for buildpack API 0.6 for the full list of changes and further details.

> ### Opt-in layer caching

Reviewer 2: ```suggestion
### Opt-in layer caching and launch.toml format changes
```

Buildpacks must now explicitly opt-in to layer re-use by setting the `launch`, `build`, and `cache` keys in `<layers>/<layer>.toml`. If buildpacks do not modify `<layers>/<layer>.toml`, the layer will behave like a temporary directory, available only to the authoring buildpack, existing for the duration of a single build - even if the buildpack in the previous build set any of these keys to `true`.

> Additionally, the `launch`, `build`, and `cache` keys are moved under a new `[types]` table in `<layers>/<layer>.toml` and removed from the top level. A Bash buildpack could write something like the following to `<layers>/<layer>.toml` in order to cache a layer:

Reviewer 2: ```suggestion
> **NOTE**: Additionally, the `launch`, `build`, and `cache` keys are moved under a new `[types]` table in `<layers>/<layer>.toml` and removed from the top level. 

A Bash buildpack could write something like the following to `<layers>/<layer>.toml` in order to cache a layer:
```

Does this work better? 🤔

Author: I like it!
